lucetc/src/compiler.rs

```diff
@@ -261,152 +255,153 @@ impl<'a> Compiler<'a> {
-                .map_err(|source| Error::FunctionTranslation {
-                    symbol: func.name.symbol().to_string(),
-                    source: Box::new(Error::from(source)),
-                })?;
-            let func_id = func.name.as_funcid().unwrap();
-            let mut traps = TrapSites::new();
-            let compiled = self
-                .clif_module
-                .define_function(func_id, &mut clif_context, &mut traps)
-                .map_err(|source| Error::FunctionDefinition {
-                    symbol: func.name.symbol().to_string(),
-                    source,
-                })?;
-
-            let size = compiled.size;
-
-            let trap_data_id = traps.write(&mut self.clif_module, func.name.symbol())?;
-
-            function_map.insert(func_id, (size, trap_data_id, traps.len()));
-        }
+        let mut decls = self.decls;
+        let codegen_context = self.codegen_context;
+        let count_instructions = self.count_instructions;
+
+        let mut function_map = self
+            .function_bodies
+            .into_par_iter()
+            .map(|(unique_func_ix, (mut validator, func_body))| {
+                let func = decls
+                    .get_func(unique_func_ix)
+                    .expect("decl exists for func body");
+                let mut func_info = FuncInfo::new(&decls, &codegen_context, count_instructions);
+                let mut clif_context = ClifContext::new();
+                clif_context.func.name = func.name.as_externalname();
+                clif_context.func.signature = func.signature.clone();
+
+                FuncTranslator::new()
+                    .translate_body(
+                        &mut validator,
+                        func_body.clone(),
+                        &mut clif_context.func,
+                        &mut func_info,
+                    )
+                    .map_err(|source| Error::FunctionTranslation {
+                        symbol: func.name.symbol().to_string(),
+                        source: Box::new(Error::from(source)),
+                    })?;
+                let func_id = func.name.as_funcid().unwrap();
+                let trap_metadata =
+                    codegen_context.compile(&mut clif_context, func_id, func.name.symbol())?;
+
+                Ok((func_id, trap_metadata))
+            })
+            .collect::<Result<HashMap<FuncId, TrapMetadata>, Error>>()?;
 
         // Now that we've defined all functions, we know what trampolines must also be created.
-        for (hostcall_name, (trampoline_id, hostcall_func_index)) in &self.trampolines {
-            synthesize_trampoline(
-                &mut self.decls,
-                &mut self.clif_module,
-                &mut function_map,
-                hostcall_name,
-                *trampoline_id,
-                *hostcall_func_index,
-            )?;
+        let trampoline_metas = codegen_context
+            .trampolines()
+            .iter()
```

Comment:
FWIW this is probably far less impactful than the main body of code but `.par_iter()` is likely imminently plausible here.

Reply:
great idea, done in aedf40349df48c2ac950d6ac526893b62e69dd40